src/adversarial/neurify.jl

```diff
@@ -0,0 +1,306 @@
+
+        LG_hat = max.(LG, 0.0) .* LΛ[i] .+ min.(LG, 0.0) .* UΛ[i]
+        UG_hat = min.(UG, 0.0) .* LΛ[i] .+ max.(UG, 0.0) .* UΛ[i]
+
+        LG, UG = interval_map(layer.weights', LG_hat, UG_hat)
+    end
+
+    # NOTE can omit this line in the paper version
+    i_max == 0 || j_max == 0 && error("Can not find valid node to split")
+
+    push!(splits, (i_max, j_max, influence_max))
+
+    return (i_max, j_max, influence_max)
+end
+
+function forward_layer(solver::Neurify, layer::Layer, input)
+    return forward_act(forward_linear(solver, input, layer), layer)
+end
+
+# Symbolic forward_linear for the first layer
+function forward_linear(solver::Neurify, input::AbstractPolytope, layer::Layer)
+    (W, b) = (layer.weights, layer.bias)
+    sym = SymbolicInterval(hcat(W, b), hcat(W, b), input)
+    LΛ = Vector{Vector{Int64}}(undef, 0)
+    UΛ = Vector{Vector{Int64}}(undef, 0)
+    r = Vector{Vector{Int64}}(undef, 0)
+    return SymbolicIntervalGradient(sym, LΛ, UΛ, r)
+end
+
+# Symbolic forward_linear
+function forward_linear(solver::Neurify, input::SymbolicIntervalGradient, layer::Layer)
+    (W, b) = (layer.weights, layer.bias)
+    output_Up = max.(W, 0) * input.sym.Up + min.(W, 0) * input.sym.Low
+    output_Low = max.(W, 0) * input.sym.Low + min.(W, 0) * input.sym.Up
+    output_Up[:, end] += b
+    output_Low[:, end] += b
+    sym = SymbolicInterval(output_Low, output_Up, input.sym.interval)
+    return SymbolicIntervalGradient(sym, input.LΛ, input.UΛ, input.r)
+end
+
+# Symbolic forward_act
+function forward_act(input::SymbolicIntervalGradient, layer::Layer{ReLU})
+    n_node, n_input = size(input.sym.Up)
+    output_Low, output_Up = copy(input.sym.Low), copy(input.sym.Up)
+    mask_lower, mask_upper = zeros(Float64, n_node), ones(Float64, n_node)
+    interval_width = zeros(Float64, n_node)
+    for i in 1:n_node
+        # Symbolic linear relaxation
+        # This is different from ReluVal
+
+        up_low, up_up = bounds(input.sym.Up[i, :], input.sym.interval)
+        low_low, low_up = bounds(input.sym.Low[i, :], input.sym.interval)
+
+        interval_width[i] = up_up - low_low
+
+        up_slop = act_gradient(up_low, up_up)
+        low_slop = act_gradient(low_low, low_up)
+
+        output_Up[i, :] = up_slop * output_Up[i, :]
+        output_Up[i, end] += up_slop * max(-up_low, 0)
+
+        output_Low[i, :] = low_slop * output_Low[i, :]
+
+        mask_lower[i], mask_upper[i] = low_slop, up_slop
+    end
+    sym = SymbolicInterval(output_Low, output_Up, input.sym.interval)
+    LΛ = push!(input.LΛ, mask_lower)
+    UΛ = push!(input.UΛ, mask_upper)
+    r = push!(input.r, interval_width)
+    return SymbolicIntervalGradient(sym, LΛ, UΛ, r)
+end
+
+function forward_act(input::SymbolicIntervalGradient, layer::Layer{Id})
+    sym = input.sym
+    n_node = size(input.sym.Up, 1)
+    LΛ = push!(input.LΛ, ones(Float64, n_node))
+    UΛ = push!(input.UΛ, ones(Float64, n_node))
+    r = push!(input.r, ones(Float64, n_node))
```

Comment:
`r` should be the width of the interval, right? Is it 1.0 here just because we can ignore `r` in the case of `Id` layers later on, so the value doesn't matter?

Reply:
This does not look correct to me. @Wei-TianHao

Reply:
Any new ideas here?